### PR TITLE
[WFAPI-3233] Use after_commit callback instead

### PR DESCRIPTION
**Pull request by gwshaw**
Systems are sufficiently fast that a PubSub message can be published, received, and processed before the database finishes the commit. We've seen this particularly with `create` where the record did not yet exist when the subscriber went to access it.

The solution is to set the change status before the transaction completes, and then decide to publish or not after the commit completes.

### CODE REVIEW

File: lib/pub_sub/railties/active_record.rb [hunk 1]
```diff
@@ -5,9 +5,14 @@ def self.included(klass)
...
 
-        klass.after_save :publish_changes, if: -> { changed? }
-        klass.after_destroy :publish_changes
-        klass.after_touch :publish_changes
+        klass.after_save { @__pub_sub_record_unchanged = !changed? }
+        # Not everything executes the `after_save` callback (like touch and destroy),
+        # so use `unless` logic so the uninitialized @__xxx_unchanged is nil, which equals false.
+        klass.after_commit :publish_changes, unless: Proc.new {
+          # return current unchanged status and reset unchanged status to false
+          x, @__pub_sub_record_unchanged = @__pub_sub_record_unchanged, false
```
Comment (ashoda) [inline lib/pub_sub/railties/active_record.rb]: Granted very concise 😄 , but this piece is a bit hard to follow. I'd lean toward being more explicit and clear as to when changes are published. This is what I have in mind as an example.

```ruby
#excplit cases that mark a record to be published
klass.after_save :__pubsub_mark_to_publish!, if: :changed?
klass.after_destroy :__pubsub_mark_to_publish!
klass.after_touch :__pubsub_mark_to_publish!
klass.after_commit :publish_changes, if: __pubsub_changes_to_publish?

def __pubsub_mark_to_publish!
  @__pubsub_changes_to_publish = true
end

def __pub_sub_marked_to_publish?
  @__pubsub_changes_to_publish
end

module InstanceMethods
  def publish_changes
    @__pubsub_changes_to_publish = false 
    # ...
  end
# ... 
```

Comment (gwshaw) [inline lib/pub_sub/railties/active_record.rb]: Yeah, I had that originally, butis is sooo many more lines of code to accomplish the same thing...

File: lib/pub_sub/railties/active_record.rb [hunk 2]
```diff
@@ -5,13 +5,25 @@ def self.included(klass)
...
-        klass.after_destroy :publish_changes
-        klass.after_touch :publish_changes
+        # excplit cases that mark a record to be published
+        klass.after_save :__pubsub_mark_to_publish!, if: :changed?
+        klass.after_destroy :__pubsub_mark_to_publish!
+        klass.after_touch :__pubsub_mark_to_publish!
+
+        klass.after_commit :publish_changes, if: :__pubsub_marked_to_publish?
+
+        def __pubsub_mark_to_publish!
```
Comment (ashoda) [inline lib/pub_sub/railties/active_record.rb]: I would move these within `InstanceMethods`

File: lib/pub_sub/railties/active_record.rb [hunk 3]
```diff
@@ -5,13 +5,25 @@ def self.included(klass)
         klass.send :include, InstanceMethods
         klass.send :extend, ClassMethods
 
-        klass.after_save :publish_changes, if: -> { changed? }
-        klass.after_destroy :publish_changes
-        klass.after_touch :publish_changes
+        # excplit cases that mark a record to be published
```
Comment (ashoda) [inline lib/pub_sub/railties/active_record.rb]: I'm not sure if we need this comment. But if we keep it, explicit is misspelled

Comment (ashoda) [inline lib/pub_sub/railties/active_record.rb]: Not sure what specs we have for this. But would be good to cover this behavior. Have you used this before? https://github.com/jdliss/shoulda-callback-matchers

**Comment by gwshaw** [comment]
Note that this change may cause unexpected breakage. By moving to `after_commit`, the number of calls to `publish` will be reduced to one per transaction rather than one per save/touch/destroy. This will still produce a message for each change, but not redundant messages for multiple publishable events on the same change. Tests that count call to `publish` may fail, and code that expects redundant messages may fail.
